domain/application/state/application.go

```diff
@@ -787,6 +787,35 @@ WHERE uuid = $applicationID.uuid
 	return life.LifeID, nil
 }
 
+// IsControllerApplication returns true when the application is the controller.
+func (st *State) IsControllerApplication(ctx context.Context, appID coreapplication.ID) (bool, error) {
+	db, err := st.DB()
+	if err != nil {
+		return false, errors.Capture(err)
+	}
+
+	controllerApp := controllerApplication{
+		ApplicationID: appID,
+	}
+	stmt, err := st.Prepare(`
+SELECT TRUE AS &controllerApplication.is_controller
+FROM application_controller
+WHERE application_uuid = $controllerApplication.application_uuid
+`, controllerApp)
+	if err != nil {
+		return false, errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, stmt, controllerApp).Get(&controllerApp)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Capture(err)
+		}
```

Comment:
The problem with this approach is that you cannot differentiate when the app doesn't exist vs the app exists but it's not a controller. In both cases you'll end up returning `false, nil` in the service. 
Maybe you could first query the application table with the given ID to provide a correct error (I think we do this in various methods where we first get the unit life or details before actually retrieving anything else).

Reply:
Despite this being unlikely, since this is set when the application is created, you are right. I'll add a check.